### ISSUE TITLE
Project maintainence

### ISSUE THREAD

**Issue by michaeljones** [michaeljones/breathe#336]
Hi everyone,

I think it has come time that I admit to myself that I will not be returning to a more active role in this project. I am very proud of the little community here, very grateful to those who have contributed and hugely happy to see people posting, helping each other and seeking to improve Breathe.

I apologise for my lack of activity. I have long felt that I can try to help out as much as I can but it is becoming harder and harder to step back in. 

It is great to see so many helping out. If there is anything I can do to help this project continue without my active involved please let me know in this thread.

Kind regards,
Michael

**Comment by michaeljones** [comment]
For the moment, I have attempted to disable posting to the Breathe google group in an effort to push people towards the community here. The group was maintained in part to let people get help without having to have a Github account but I feel that not enough people monitor the group for it to be useful.

**Comment by Andne** [comment]
I can think of two things that should help keep the project moving forward:
1. There's some open pull requests that people would like included, some clear guidelines on what you think should be done before merging PR's in general (some of us have those permissions now I think) in order to maintain project quality would be nice.  We don't really know what your criteria was.
2. How do new releases get created once we get some of the open pull requests merged?

Another possible direction entirely, should ownership of breathe be transferred?  The first possibility that comes to my mind is to assign it to the sphinx maintainers similar to some of the other extensions they already have.  I think it may be best off as a separate project over there (kind of like their websupport extension) if this is the right decision, but seems like a possibility that would allow for guidance on continued development.

**Comment by SylvainCorlay** [comment]
Hi, I just wanted to add to the thread that this project is of tremendous help for us, and also - I think - follows the right approach to the problem it is solving. Kudos to @michaeljones for pulling this off and sorry to see that you don't have time anymore for working on it!

It would be awesome if another maintainer that is knowledgeable about the code base steps up so that the project continues further!

**Comment by vitaut** [comment]
Hi Michael,

Thanks for creating this project and all your amazing work! Sorry that I haven't been able to help with maintenance lately. I've been struggling to find time even for my main open-source project, but fortunately had some help from the community. I hope that someone will be willing to step in and help with the maintenance here as well.

Good luck!
Victor

**Comment by SylvainCorlay** [comment]
@vitaut do you have commit rights on the project?

**Comment by michaeljones** [comment]
Thank you @vitaut. Thank you for the kind words and for the incredible work you've put into this project. You've kept it alive and well long after I had slowed down on it and your contributions help to motivate a few more of mine. 

I have always been impressed that you found time to contribute when you already had a much more significant project to look after. Thank you for your efforts.

**Comment by michaeljones** [comment]
Regarding @Andne comments. I wouldn't worry about my criteria for contributions at this point. I think the most important thing is that the people who are interested in the project and its maintenance do their best to get at least one or two other people to agree to changes and then to move forward as best they can. I am happy to add more people as collaborators to help with that. Other than that, I'd try to keep the scope of the project contained as expanding it is only a greater maintenance burden.
 
With regards to releases, I am happy to give interested people access to the Breathe PyPI administration so that they can do releases as well when necessary.

As far moving the repository, I am very happy for it to move into new ownership. The sphinx maintainers group might be a good fit but I feel organisations like them can already have too much to deal with. Still, I'm happy to hear thoughts and ideas.

**Comment by svenevs** [comment]
How many people do you think would be a reasonable number of "minimum core maintainers"?  I would be interested in helping maintain things, but I'm transitioning and have no idea what kind of free time I will actually have until ~December.

I think if it were possible to get somewhere between 3-5 people to commit to best-faith-effort `breathe` maintenance then bug-fixes and the like could at least still be merged in.  The GitHub [PR reviews](https://help.github.com/articles/approving-a-pull-request-with-required-reviews/) are pretty nice, it could just get set as needs 2 reviews.

How to use breathe is generally very well documented, but the code itself is pretty easy to get lost in.  I know a lot of that is from the api generator thing, but explaining things like how `breathe.parser.index` and `breathe.parser.compound` differ, how the actual `breathe.__init__` works (aka how setting the `breathe_projects` in `conf.py` indirectly configures this).  Just kind of the high level outline would be useful as a reference for future maintainers, as well as potential future contributors.

**Comment by michaeljones** [comment]
Thanks for raising that @svenevs. In regards to the documentation, I very happy to try to contribute some more to smooth the transition for anyone who would like to help out. I'll try to work on those points you've raised and we'll see if there is more to talk about.

**Comment by SylvainCorlay** [comment]
@vitaut @michaeljones 

@JohanMabille and myself are both involved in large open-source projects (jupyter & co), and rely on breathe for a number of other things [xtensor](https://xtensor.readthedocs.io/),  [xtensor-python](https://xtensor-python.readthedocs.io/),  [xtensor-julia](https://xtensor-julia.readthedocs.io/), [xeus](https://xeus.readthedocs.io/),  [xproperty](https://xproperty.readthedocs.io/). I have some experience working on sphinx extensions but do not know the internals of breathe well.

If you are interested, we can help with keeping the project on life support by doing releases and basic housekeeping.

For example, the current version of breathe does not work with the latest stable release of sphinx, while the bug is fixed in master. A release of sphinx would help a lot of the users who rely on RTD for example.

If you are interested in this, let us know.

**Comment by michaeljones** [comment]
Hi @SylvainCorlay, thank you for volunteering to help out. Very kind of you. I'd be very happy to grant you access for contributions & releases if you'd like. Most things can be posted here but feel free to drop me an email on m.pricejones@gmail.com if you would like to pick up anything off thread. I guess I would need your PyPI username if nothing else in order to allow you to do releases.

**Comment by SylvainCorlay** [comment]
Great! I will shoot you an email with all the information!

**Comment by michaeljones** [comment]
For the benefit of others reading this, I have added you both to Breathe on PyPI and responded to your email.

**Comment by SylvainCorlay** [comment]
Great, thanks!

**Comment by svenevs** [comment]
I may have misunderstood, was there to be a quick release from `master` (or cherry-picked [this guy](https://github.com/michaeljones/breathe/commit/03ea574ef6034019c2e928c5733b24e64ec8b651)) sometime soon?  Otherwise I'd like to submit a PR with hotfix directions to bypass this on RTD (either choosing older sphinx, or installing breathe from source rather than `pip`) so that people can "just know" how to bypass :wink:

If the latter, PR against the `README.rst` or `documentation/source/index.rst`?

**Comment by SylvainCorlay** [comment]
There should be a release in the next few days. I only need commit rights to tag one, then we will release it on conda and pypi.

FYI, the package on conda-forge already has the fix, hence for RTD, all you need is a `readthedocs.yml` at the root of the repository, just like [here](https://github.com/QuantStack/xtensor/blob/master/readthedocs.yml) specifying the location of the `environment.yml` where the conda requirements are specified.

This one may contain

name: xtensor-docs

```yml
channels:
  - conda-forge

dependencies:
  - breathe
```

**Comment by svenevs** [comment]
Woot!  Thanks so much!!!

Yes, it seems I've fallen behind on the latest and greatest RTD features -- I didn't even know about the [rtd yaml](http://docs.readthedocs.io/en/latest/yaml-config.html).  So my would-be hack-around instructions would have been much more complicated than needed!

**Comment by SylvainCorlay** [comment]
Hi all, I have release breathe 4.7.0 on pypi and conda-forge, and added the related information to the changelog. I have built a number of projects docs with this version and the latest version of sphinx on RTD and everything seems to work as expected.

**Comment by melvinvermeeren** [comment]
Hi, I would also like to help out maintaining this. At the very least I will review pull requests and release new versions when those changes have been reviewed and merged. When time allows it I will also try to fix some relatively simple issues.

@SylvainCorlay I contacted Michael already and we came to the conclusion that posting here would be useful so current maintainers will know what's going on. Do you approve of me helping out with the project?

**Comment by SylvainCorlay** [comment]
Welcome @melvinvermeeren !

I think the more the merrier. I have mostly been doing housekeeping and we pushed out new versions with minor bug fixes so far.

There is also @ax3l who has been doing some more in-depth work on the memory consumption issues for large projects.

**Comment by melvinvermeeren** [comment]
Closing for now since I should be able to maintain this for an extended period of time. When doc is updated (#357) I'll add some mention that we are looking for contributors.

If any of you still remember details about the Breathe architecture then I would appreciate having those details. Having diagrams of high level components and main program flow would help a lot for new contributors I think. http://plantuml.com/ is probably the easiest way to add them, there is a sphinx extension available so the diagrams' source can be inlined in the documentation sources. 

Thanks for all the work so far!

**Comment by melvinvermeeren** [comment]
I believe commenting here is the easiest way to hopefully reach most (ex-)maintainers.

I urge everyone to have a look at #383 regarding future management and automation of the project.

Thank you.

**Comment by svenevs** [comment]
I would like to request write access. I'm on the sixth iteration of the new parser using `lxml` and think I'll finally make it out of Mordor this time.

Being able to open a branch / pull request on this repository will make it easier for current maintainers to test it, rather than a fork.

I will not be touching `master`, but long term I will be contributing more work for things like `using`, template specializations, and some bypasses for `exhale` :wink:

**Comment by michaeljones** [comment]
Sounds good to me. I'm happy to add your as a contributor to aid this change but, yes, please make sure you coordinate with @melvinvermeeren over any changes to be merged into master.
